### PR TITLE
[Web Portal] Reduce compile time; Use webpack-dev-server for development

### PR DESCRIPTION
**Pull request by sunqinzheng**
- Upgrade webpack -> 4.x
- ExtractTextPlugin -> MiniCssExtractPlugin 
- UglifyJsPlugin -> TerserPlugin 
- Use beta version html-webpack-plugin
  - https://github.com/jantimon/html-webpack-plugin/issues/724#issuecomment-419885840

### CODE REVIEW

**Comment by Gerhut** [comment]
Could you please provide the compile time before & after reducing?

**Comment by sunqinzheng** [comment]
new
prod-full-build: 100.4086647s
dev-full-build: 32.3711059s
dev-file-change: 1.21s
old
prod-full-build: 102.9010536s
dev-full-build: 51.5960864s
dev-file-change: 7.86s

File: src/webportal/config/webpack.common.js [hunk 1]
```diff
@@ -17,67 +17,68 @@
...
-    hardware: './src/app/cluster-view/hardware/hardware.component.js',
-    hardwareDetail: './src/app/cluster-view/hardware/hardware-detail.component.js',
-    k8s: './src/app/cluster-view/k8s/k8s.component.js',
-    docs: './src/app/job/job-docs/job-docs.component.js',
-    jobSubmit: './src/app/marketplace/job-submit/job-submit.component.js',
-    changeGitHubPAT: './src/app/user/change-github-pat/change-github-pat.component.js',
-    howToConfigGitHubPAT: './src/app/user/how-to-config-github-pat/how-to-config-github-pat.component.js',
-    plugin: './src/app/plugin/plugin.component.js',
-
+    'index': './src/app/index.js',
```
Comment (debuggy) [inline src/webportal/config/webpack.common.js]: why use single queto in ```index```? I saw the [official doc](https://webpack.js.org/concepts/entry-points) and did not see this convention.

File: src/webportal/package.json [hunk 2]
```diff
@@ -84,19 +85,21 @@
...
   },
   "scripts": {
     "lint": "eslint --ext .js --ext .jsx src",
     "test": "npm run lint",
     "clean": "rimraf dist",
     "prebuild": "npm run clean && node config/preprocess.js",
-    "build": "webpack --config webpack.config.js --progress --profile",
-    "build:dev": "npm run build -- --debug --devtool eval --output-pathinfo --watch",
-    "prestart": "envsub --env-file .env --system src/app/env.js.template dist/scripts/env.js",
+    "build": "webpack --config webpack.config.js --progress --profile -p",
```
Comment (Gerhut) [inline src/webportal/package.json]: I suggest provide the full, expanded arguments instead of `-p` for better reading.

Comment (Gerhut) [inline src/webportal/package.json]: If `webpack.config.js` is the default configuration path we could omit it like `dev` script.

The difference between `webpack --config` in `build` and `webpack-dev-server` in `dev` may lead confusing.

File: src/webportal/config/webpack.common.js [hunk 3]
```diff
@@ -17,67 +17,68 @@
...
-  removeComments: true,
-  removeEmptyAttributes: true,
-  removeTagWhitespace: true
-};
+function generateHtml(opt) {
+  return new HtmlWebpackPlugin(Object.assign({
+    title: title,
+    version: version,
+    template: './src/app/layout/layout.component.ejs',
+    minify: {
```
Comment (Gerhut) [inline src/webportal/config/webpack.common.js]: Prevent minification during development may increase the build speed.

File: src/webportal/package.json (hunk 2 above)
Comment (debuggy) [reply src/webportal/package.json]: Agree, maybe we could provide a ```webpack.dev.config.js``` for ```yarn dev``` and a ```webpack.prod.config.js``` for ```yarn build```.

File: src/webportal/config/webpack.common.js [hunk 4]
```diff
@@ -173,233 +163,146 @@ const config = (env, argv) => ({
...
+    // required by ejs loader
     new webpack.ProvidePlugin({
-      _: 'underscore'
+      _: 'lodash',
     }),
     new webpack.ProvidePlugin({
-      $: 'jquery',
-      jQuery: 'jquery',
-      'window.jQuery': 'jquery'
+      '$': 'jquery',
```
[the remark this answers is not in the document]
Comment (Gerhut) [reply src/webportal/config/webpack.common.js]: I prefer add quote for unify keys in same object.

File: src/webportal/config/webpack.common.js (hunk 3 above)
Comment (sunqinzheng) [reply src/webportal/config/webpack.common.js]: JS minify is disabled in dev mode. And I think html minify is fast

File: src/webportal/config/webpack.common.js (hunk 1 above)
Comment (sunqinzheng) [reply src/webportal/config/webpack.common.js]: https://eslint.org/docs/rules/quote-props
This rule is enabled in our eslint config file.

File: src/webportal/package.json (hunk 2 above)
Comment (sunqinzheng) [reply src/webportal/package.json]: > 
> 
> Agree, maybe we could provide a `webpack.dev.config.js` for `yarn dev` and a `webpack.prod.config.js` for `yarn build` with different optimizations.

https://webpack.js.org/api/cli/#shortcuts
I think webpack's "-d" and "-p" flag is enough. The shortcuts includes source map and minify options.

File: src/webportal/config/webpack.common.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (debuggy) [reply src/webportal/config/webpack.common.js]: that's fine

File: src/webportal/config/webpack.common.js (hunk 1 above)
Comment (debuggy) [reply src/webportal/config/webpack.common.js]: got it

File: src/webportal/package.json (hunk 2 above)
Comment (sunqinzheng) [reply src/webportal/package.json]: fixed

Comment (sunqinzheng) [reply src/webportal/package.json]: fixed.
use webpack-cli's "--mode" option.

**Comment by abuccts** [comment]
@sunqinzheng 
It seems this PR breaks dev build: `npm run dev` then `npm start` cannot start webportal.

**Comment by sunqinzheng** [comment]
Only ```npm run dev``` is needed. Pages will be served by webpack-dev-server (https://github.com/webpack/webpack-dev-server) at localhost:8080 (the link will be shown in terminal).

**Comment by sunqinzheng** [comment]
#2421

**Pull request merged**